### ISSUE TITLE
[AppGate] Enable auto-undelegation from Unstaking Gateways

### ISSUE THREAD

**Issue by h5law** [pokt-network/poktroll#111]
## Objective

This issue requires numerous things:
- Applications being automatically undelegated from unstaking gateways
- Applications keeping delegated gateways by their addresses not public keys

Currently we cannot undelegate from an unstaked gateway as we require the `gatewayAddress`->`gatewayPubKey` step. This relies on the `accountKeeper` which can only query information for staked actors.

We should instead store the delegated gateways by their address strings. A side effect of converting the list of delegated gateways the application maintains is that it is able to remove the delegated gateway without having to query this on chain data. However as mentioned in this comment https://github.com/pokt-network/poktroll/pull/90#discussion_r1375222403 if applications are not automatically undelegated from unstaking gateways we will be unable to build the ring for them off-chain

## Origin Document

`x/application/keeper/msg_server_undelegate_from_gateway.go`
```go
	// Check if the gateway is staked
	// TODO(@h5law): Look into using addresses instead of public keys
	if _, found := k.gatewayKeeper.GetGateway(ctx, msg.GatewayAddress); !found {
		logger.Info("Gateway not found with address [%s]", msg.GatewayAddress)
		return nil, sdkerrors.Wrapf(types.ErrAppGatewayNotFound, "gateway not found with address: %s", msg.GatewayAddress)
	}
```
`x/application/keeper/msg_server_undelegate_from_gateway_test.go`
```go
// TODO_TECHDEBT(@h5law): This should not be a cause of failure for unstaking
func TestMsgServer_UndelegateFromGateway_FailGatewayNotStaked(t *testing.T) {
```
## Goals

- Research and implement the usage of `depinject` so that both the `application` and `gateway` modules can communicate with one another
- Enable auto undelegation for unstaking gateways
- Enable the usage of address strings instead of public keys in the applications delegated gateways section
- In turn enable the creation of rings seamlessly off-chain

## Deliverables

- [ ] A **single PR** that contains:
  - [ ] Address strings instead of public keys stored on chain for delegated gateways
  - [ ] `depinject` and communication between the application and gateway modules for auto undelegation
  - [ ] storage of all applications that a gateway has been delegated to (for auto undelegation purposes)
  - [ ] extensive tests covering this behaviour

## Non-goals / Non-deliverables

- Altering any (un)delegation logic outside of the scope of using address strings
- Altering the gateway unstaking logic outside of appending some undelegation logic

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
- [ ] **Makefile**: Add new targets to the Makefile to make the new functionality easier to use.
- [ ] **Documentation**: Update architectural or development READMEs; use [mermaid](https://mermaid-js.github.io/mermaid/) diagrams where appropriate.

---

**Creator**: @h5law
**Co-Owners**: @Olshansk

**Comment by Olshansk** [comment]
Note: Going to reply in a few different messages to keep unrelated context separately.

> Applications being automatically undelegated from unstaking gateways

NIT clarification on wording: `Applications automatically undelegating unstaking gateways` since applications are the onces doing the action of delegation/undelegation and gateways the ones that are delegated to/from.

**Comment by Olshansk** [comment]
>  This relies on the accountKeeper which can only query information for staked actors.

I want to challenge this statement and understand why only staked actors can be queried?

---

My understanding is that all accounts exist in perpetuity but only become available in the store once they have been interacted with. For example, even as simple as sending them funds.

This is how we got these tests to pass in `testutil/network/network.go`.

```
func InitAccount(t *testing.T, net *Network, addr sdk.AccAddress) {
	t.Helper()
	val := net.Validators[0]
	ctx := val.ClientCtx
	args := []string{
		fmt.Sprintf("--%s=%s", flags.FlagFrom, val.Address.String()),
		fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
		fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastSync),
		fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(net.Config.BondDenom, sdkmath.NewInt(10))).String()),
	}
	amount := sdk.NewCoins(sdk.NewCoin("stake", sdkmath.NewInt(200)))
	_, err := clitestutil.MsgSendExec(ctx, val.Address, addr, amount, args...)
	require.NoError(t, err)
}
```

Furthermore, ChatGPT seems to have reaffirmed this understanding.

![Screenshot 2023-10-30 at 11 26 14 AM](https://github.com/pokt-network/poktroll/assets/1892194/a991b54c-41e9-4652-88f8-d5e5928af0e6)

**Comment by Olshansk** [comment]
> However as mentioned in this comment https://github.com/pokt-network/poktroll/pull/90#discussion_r1375222403 if applications are not automatically undelegated from unstaking gateways we will be unable to build the ring for them off-chain

![Screenshot 2023-10-30 at 11 28 59 AM](https://github.com/pokt-network/poktroll/assets/1892194/ca2aa395-653d-40df-bf64-3f73c6fe334f)

Going to push back on this statement:

> retrieval of the gateway's public key from their address potentially impossible

My understanding (per the previous comment) is that once an actor has had any sort of on-chain interaction, it'll be available via the AccountKeeper in perpetuity. Happy to re-evaluate if you can show that's not the case.

**Comment by Olshansk** [comment]
> storage of all applications that a gateway has been delegated to (for auto undelegation purposes)

Regardless of whether we go with auto-undelegation or not, I do not think it's a good idea to replicate the storage because:
1. It creates TWO SOUCES OF TRUTH for a single piece of data that MUST be kept in sync`
2. It breaks some rules of `OOP` (e.g. in a family tree, children don't usually have backlinks to their parents)
3. Apps will delegate to few gateways, but gateways will have many (e.g. tens of thousands) of apps delegating to them. This can become a very big & expensive structure.

----

If we do go with auto-undelegation, I'd prefer to keep this data the application keeper, but simply add an additional layer of indexing. Below is not a full throughout suggestion, but just a rough idea.

Here are some components of indexing, storing & accessing applications.

```go
ApplicationKeyPrefix = "Application/value/"

// ...

func KeyPrefix(p string) []byte {
	return []byte(p)
}

// ...

store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ApplicationKeyPrefix))

// ...

// SetApplication set a specific application in the store from its index
func (k Keeper) SetApplication(ctx sdk.Context, application types.Application) {
	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.ApplicationKeyPrefix))
	b := k.cdc.MustMarshal(&application)
	store.Set(types.ApplicationKey(
		application.Address,
	), b)
}

```

For example, what if were to update `SetApplication` to also update a store that uses the prefix `Application/gateway/` that maps `/gatewayAddr/appAddr -> bool` that is updated during delegation/undelegation. That way:

1. We can use the `gatewayAddr` prefix to find all apps to Unstake
2. It's an atomic operation with one source of truth.

**Comment by Olshansk** [comment]
> In turn enable the creation of rings seamlessly off-chain

Ignoring the entire previous comment. I still don't fully understand why this is not possible.

**Comment by red-0ne** [comment]
Closed as per #692
